File: ios/CameraModuleViewManager.m
#import "React/RCTViewManager.h"

@interface RCT_EXTERN_MODULE(CameraModuleViewManager, RCTViewManager)

RCT_EXPORT_VIEW_PROPERTY(device, NSString)
RCT_EXPORT_VIEW_PROPERTY(flash, NSString)
RCT_EXPORT_VIEW_PROPERTY(torch, NSString)
RCT_EXPORT_VIEW_PROPERTY(deviceType, NSString)
RCT_EXPORT_VIEW_PROPERTY(zoom, CGFloat)

// Camera View Functions
RCT_EXTERN_METHOD(start:(nonnull NSNumber *)node);
RCT_EXTERN_METHOD(stop:(nonnull NSNumber *)node);
RCT_EXTERN_METHOD(startRecording:(nonnull NSNumber *)node options:(NSDictionary *)options onRecordCallback:(RCTResponseSenderBlock)onRecordCallback);
RCT_EXTERN_METHOD(pauseRecording:(nonnull NSNumber *)node);
RCT_EXTERN_METHOD(resumeRecording:(nonnull NSNumber *)node);
RCT_EXTERN_METHOD(stopRecording:(nonnull NSNumber *)node);
RCT_EXTERN_METHOD(takePhoto:(nonnull NSNumber *)node options:(NSDictionary *)options resolve:(RCTPromiseResolveBlock)resolve reject:(RCTPromiseRejectBlock)reject);
RCT_EXTERN_METHOD(focus:(nonnull NSNumber *)node point:(NSDictionary *)point resolve:(RCTPromiseResolveBlock)resolve reject:(RCTPromiseRejectBlock)reject);
RCT_EXTERN_METHOD(getDevices:(nonnull NSNumber *)node position:(NSString *)position devicesList:(RCTResponseSenderBlock)onRecordCallback);

@end

File: ios/CameraModuleViewManager.swift
import NextLevel
import AVFoundation
import UIKit
import Foundation
import React

@objc(CameraModuleViewManager)
class CameraModuleViewManager: RCTViewManager {
    
    override func view() -> UIView! {
        return CameraView()
    }
    
    override var methodQueue: DispatchQueue! {
        return DispatchQueue.main
    }
    
    @objc
    final func start(_ node: NSNumber) {
        let component = getCameraView(withTag: node)
        component.start()
    }
    
    
    @objc
    final func stop(_ node: NSNumber) {
        let component = getCameraView(withTag: node)
        component.stop()
    }
    
    @objc
    final func startRecording(_ node: NSNumber, options: NSDictionary, onRecordCallback: @escaping RCTResponseSenderBlock) {
        let component = getCameraView(withTag: node)
        component.startCapture(options: options, callback: onRecordCallback)
    }
    
    @objc
    final func stopRecording(_ node: NSNumber) {
        let component = getCameraView(withTag: node)
        component.endCapture()
    }
    
    @objc
    final func pauseRecording(_ node: NSNumber) {
        let component = getCameraView(withTag: node)
        component.pauseCapture()
    }
    
    @objc
    final func getDevices(_ node: NSNumber, position: String, devicesList: @escaping RCTResponseSenderBlock) {
        let component = getCameraView(withTag: node)
        let callback = Callback(devicesList);
        callback.resolve(component.getSupportedDevices(position: position))
    }
    
    private func getCameraView(withTag tag: NSNumber) -> CameraView {
        return self.bridge.uiManager.view(forReactTag: tag) as! CameraView
    }
}

extension UIView {
    var parentViewController: UIViewController? {
        var parentResponder: UIResponder? = self
        while parentResponder != nil {
            parentResponder = parentResponder!.next
            if let viewController = parentResponder as? UIViewController {
                return viewController
            }
        }
        return nil
    }
}
